### PR TITLE
Deploy: add sources index to requirements.txt file before deploy

### PR DESCRIPTION
**Pull request by DharmeshPandav**
https://github.com/scrapinghub/shub/issues/381

### CODE REVIEW

**Comment by phrfpeixoto** [comment]
Could you please add test cases to expose the issue?

File: requirements.txt [hunk 1]
```diff
@@ -17,6 +17,7 @@ requests==2.22.0
 retrying==1.3.3
 scrapinghub==2.2.1
 six==1.10.0
+toml==0.10.0
```
Comment (vshlapakov) [inline requirements.txt:20]: pipenv isn't required to run the library, I don't think we need to add `toml` for the specific case

File: README.rst [hunk 2]
```diff
@@ -9,7 +9,7 @@ without leaving the command line.
 Requirements
 ------------
 
-* Python 2.7+ or Python 3.3+
+* Python 2.7+ or Python 3.3+ <= 3.6
```
Comment (vshlapakov) [inline README.rst:12]: I'm curious why we need to restrain Python version here, is there a reason?

File: shub/deploy.py [hunk 3]
```diff
@@ -189,8 +192,16 @@ def _get_pipfile_requirements():
         # Scrapy Cloud also doesn't support editable packages
         if 'editable' in v:
             del v['editable']
-    return open(convert_deps_to_pip(deps), 'rb')
-
+    return open(_add_sources(convert_deps_to_pip(deps), _sources=sources.encode()), 'rb')
+
+def _add_sources(_reqs_file, _sources):
+    tmp = tempfile.NamedTemporaryFile(delete=False, suffix="-requirements.txt")
```
Comment (vshlapakov) [inline shub/deploy.py:198]: The temporary file should be deleted in the end of deploy step, I believe we shouldn't rely on an OS cleanup policy. As an option we could use a tracked named temporary file like it's used [here](https://github.com/pypa/pipenv/blob/dad1c2031e239df7f882c6ee38d5ce1fe2ca1a40/pipenv/utils.py#L1441)

File: shub/deploy.py [hunk 4]
```diff
@@ -171,12 +171,15 @@ def _is_pipfile(name):
...
 def _get_pipfile_requirements():
     try:
-        from pipenv.utils import convert_deps_to_pip
+        from pipenv.utils import convert_deps_to_pip, prepare_pip_source_args
     except ImportError:
         raise ImportError('You need pipenv installed to deploy with Pipfile')
     try:
         with open('Pipfile.lock') as f:
-            deps = json.load(f)['default']
+            pipefile = json.load(f)
```
Comment (vshlapakov) [inline shub/deploy.py:179]: Could you please rename the variable as `pipfile`?

File: README.rst (hunk 2 above)
Comment (DharmeshPandav) [reply README.rst:12]: tried pytest with python 3.7 and tests were failing 
all tests were failing to initialize 

plus travis, appvoyer and tox file, all runs on version less than 3.7

may be one for separate PR
error log
```
=================================================================== ERRORS ===================================================================
___________________________________________________ ERROR collecting tests/test_deploy.py ____________________________________________________
../../venv/shub3/lib/python3.7/site-packages/_pytest/python.py:177: in pytest_pycollect_makeitem
    raise StopIteration
E   StopIteration

The above exception was the direct cause of the following exception:
../../venv/shub3/lib/python3.7/site-packages/_pytest/runner.py:163: in __init__
    self.result = func()
../../venv/shub3/lib/python3.7/site-packages/_pytest/main.py:476: in _memocollect
    return self._memoizedcall('_collected', lambda: list(self.collect()))
../../venv/shub3/lib/python3.7/site-packages/_pytest/main.py:347: in _memoizedcall
    res = function()
../../venv/shub3/lib/python3.7/site-packages/_pytest/main.py:476: in <lambda>
    return self._memoizedcall('_collected', lambda: list(self.collect()))
../../venv/shub3/lib/python3.7/site-packages/_pytest/python.py:412: in collect
    return super(Module, self).collect()
../../venv/shub3/lib/python3.7/site-packages/_pytest/python.py:326: in collect
    res = self.makeitem(name, obj)
../../venv/shub3/lib/python3.7/site-packages/_pytest/python.py:338: in makeitem
    collector=self, name=name, obj=obj)
../../venv/shub3/lib/python3.7/site-packages/_pytest/vendored_packages/pluggy.py:745: in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
../../venv/shub3/lib/python3.7/site-packages/_pytest/vendored_packages/pluggy.py:339: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
../../venv/shub3/lib/python3.7/site-packages/_pytest/vendored_packages/pluggy.py:334: in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
../../venv/shub3/lib/python3.7/site-packages/_pytest/vendored_packages/pluggy.py:613: in execute
    return _wrapped_call(hook_impl.function(*args), self.execute)
../../venv/shub3/lib/python3.7/site-packages/_pytest/vendored_packages/pluggy.py:250: in _wrapped_call
    wrap_controller.send(call_outcome)
E   RuntimeError: generator raised StopIteration
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
========================================================== 1 error in 0.68 seconds ===========================================================
```

File: tests/requirements.txt [hunk 5]
```diff
@@ -10,7 +10,7 @@ flake8==3.3.0
 mccabe==0.6.1             # via flake8
 mock==2.0.0
 pbr==1.10.0               # via mock
-pipenv==11.10.1
+pipenv==2018.11.26
```
Comment (DharmeshPandav) [inline tests/requirements.txt:13]: https://github.com/pypa/pipenv/issues/3469

Comment (DharmeshPandav) [inline tests/requirements.txt:13]: https://github.com/pypa/pipenv/issues/3469

File: README.rst (hunk 2 above)
Comment (DharmeshPandav) [reply README.rst:12]: pytest is outdated as well

Comment (DharmeshPandav) [reply README.rst:12]: many dependencies are outdated and either throwing error or warning
pytest
pipenv
click
scrapinghub ( throwing abc meta warning)

**Comment by DharmeshPandav** [comment]
@vshlapakov I am of the view the dependencies upgrade should be done out side of this PR , separately

will implement file delete instead of using pipenv.vendon create_tracked_tempfile to close this pr

File: README.rst (hunk 2 above)
Comment (vshlapakov) [reply README.rst:12]: I see, thanks for checking. You're right, many dependencies are outdated, but that should be addressed separately, let's not mix the different things together and focus on the sources thing here

File: requirements.txt [hunk 6]
```diff
@@ -7,7 +7,7 @@
 backports.ssl-match-hostname==3.5.0.1  # via docker-py
 certifi==2019.6.16        # via requests
 chardet==3.0.4            # via requests
-click==6.6
+click==7.1.2
```
Comment (vshlapakov) [inline requirements.txt:10]: I don't think it's related anyhow with the pipenv support, is it? If not, let's better update it separately

**Comment by DharmeshPandav** [comment]
have done some code cleanup

passing along the temp-dir that's created while building the egg to store  new requirements.txt file with source
it will be deleted along with any egg in tmp directory in finally https://github.com/scrapinghub/shub/blob/1fdc75c2de7f838cbc61c953593f942f12c6046f/shub/deploy.py#L119

**Comment by DharmeshPandav** [comment]
> Looks awesome, thanks a lot! 👏

Thanks 

We are using scrapy cloud auto deploy where I am migrating out projects to use `pipfile` deploy

any timeline on when this change will be deployed internally and used by scrapy cloud auto deploy?

**Comment by vshlapakov** [comment]
I hate to disappoint you but the logic isn't used as-is in Scrapy Cloud. I'll try to do my best to port the changes there but can't provide any estimates for now

**Comment by DharmeshPandav** [comment]
> I hate to disappoint you but the logic isn't used as-is in Scrapy Cloud. I'll try to do my best to port the changes there but can't provide any estimates for now

sounds great. good to hear that it is on the card and will be worked on/ported in near future.
 
Appreciation message below

We are a satisfied, happy and a long time SH user since early 2015

**Pull request merged**